File: packages/python-sdk/harness/chaos.py
"""
Chaos harness for Python SDK orchestration testing.

Makes real (or realistically delayed) HTTP calls so timing and
cancellation behavior are tested under real network conditions.

Usage:
    python -m harness.chaos --iterations 20 --policy enforce
"""

import argparse
import random
import time
import uuid
import sys
import os
from typing import Any, Callable, Dict, List, Optional, Tuple
from dataclasses import dataclass, field

sys.path.insert(0, os.path.join(os.path.dirname(__file__), "..", "src"))

from fivetoone import (
    FiveToOneClient,
    RuntimePolicy,
    CancellationToken,
    LatencyBudgetExceeded,
)

try:
    import httpx
except ImportError:
    httpx = None

# ─── Failure modes ────────────────────────────────────────────────
FailureMode = str  # "success" | "timeout" | "429" | "500" | "cost_overrun"

@dataclass
class ChaosConfig:
    iterations: int = 20
    failure_mix: Dict[str, float] = field(default_factory=lambda: {
        "success": 0.4,
        "timeout": 0.15,
        "429": 0.15,
        "500": 0.15,
        "cost_overrun": 0.15,
    })
    primary_model: str = "gpt-4o"
    fallback_models: List[str] = field(default_factory=lambda: ["claude-sonnet-4-20250514", "gpt-4o-mini"])
    latency_budget_ms: int = 5000
    max_cost_per_run: float = 0.05
    max_cost_per_day: float = 0.50
    base_url: str = "https://api.openai.com"
    api_key: str = ""
    run_id: str = ""
    host: str = "http://localhost:3000"

    def __post_init__(self):
        if not self.run_id:
            self.run_id = f"chaos_{int(time.time())}"
        if not self.api_key:
            self.api_key = os.environ.get("OPENAI_API_KEY", "")


@dataclass
class IterationResult:
    iteration: int
    model: str
    failure_mode: str
    success: bool
    latency_ms: int
    error: Optional[str] = None
    blocked: Optional[bool] = None
    fallback_reason: Optional[str] = None
    fail_open: Optional[bool] = None


# ─── Chaos call factory ───────────────────────────────────────────
def create_chaos_call(config: ChaosConfig, mode: str) -> Callable[[str, CancellationToken], Any]:
    def call_fn(model: str, token: CancellationToken) -> Any:
        if mode == "success":
            token.check()  # Check before starting
            if httpx:
                client = httpx.Client(timeout=30)
                try:
                    resp = client.post(
                        f"{config.base_url}/v1/chat/completions",
                        headers={
                            "Content-Type": "application/json",
                            "Authorization": f"Bearer {config.api_key}",
                        },
                        json={
                            "model": model,
                            "messages": [{"role": "user", "content": "Say exactly: chaos test ok"}],
                            "max_tokens": 10,
                        },
                    )
                    if resp.status_code != 200:
                        raise Exception(f"HTTP {resp.status_code}: {resp.text}")
                    return resp.json()
                finally:
                    client.close()
            else:
                import requests as req
                resp = req.post(
                    f"{config.base_url}/v1/chat/completions",
                    headers={
                        "Content-Type": "application/json",
                        "Authorization": f"Bearer {config.api_key}",
                    },
                    json={
                        "model": model,
                        "messages": [{"role": "user", "content": "Say exactly: chaos test ok"}],
                        "max_tokens": 10,
                    },
                    timeout=30,
                )
                if resp.status_code != 200:
                    raise Exception(f"HTTP {resp.status_code}: {resp.text}")
                return resp.json()

        elif mode == "timeout":
            sleep_ms = config.latency_budget_ms + 2000
            start = time.time()
            while (time.time() - start) * 1000 < sleep_ms:
                token.check()
                time.sleep(0.1)
            return {"content": "should not reach here"}

        elif mode == "429":
            err = Exception("429 Rate limit exceeded")
            err.status_code = 429
            raise err

        elif mode == "500":
            err = Exception("500 Internal Server Error")
            err.status_code = 500
            raise err

        elif mode == "cost_overrun":
            return {"content": "x" * 10000, "_metadata": {"estimatedCostUsd": 1.0}}

        else:
            raise ValueError(f"Unknown failure mode: {mode}")

    return call_fn


def pick_failure_mode(mix: Dict[str, float]) -> str:
    modes = list(mix.keys())
    weights = list(mix.values())
    return random.choices(modes, weights=weights, k=1)[0]


# ─── Run harness ──────────────────────────────────────────────────
def run_chaos_harness(config: ChaosConfig) -> List[IterationResult]:
    client = FiveToOneClient(api_key=config.api_key, host=config.host)
    results: List[IterationResult] = []

    for i in range(config.iterations):
        mode = pick_failure_mode(config.failure_mix)
        all_models = [config.primary_model] + config.fallback_models
        policy = RuntimePolicy(
            enforcement_mode="enforce",
            fallback_chain=all_models,
            max_cost_per_run=config.max_cost_per_run,
            max_cost_per_day=config.max_cost_per_day,
            latency_budget_ms=config.latency_budget_ms,
            max_attempts=2,
            backoff_ms=500,
            backoff_multiplier=2,
            retryable_errors=["429", "5xx", "rate_limit", "timeout"],
        )

        calculate_cost = (lambda result, model: 1.0) if mode == "cost_overrun" else None

        start = time.time()
        success = False
        error_msg = None

        try:
            client.orchestrate(
                input_data={"chaosIteration": i, "failureMode": mode},
                policy=policy,
                call=create_chaos_call(config, mode),
                model=config.primary_model,
                calculate_cost=calculate_cost,
                run_id=f"{config.run_id}_iter{i}",
            )
            success = True
        except Exception as e:
            error_msg = str(e)

        latency_ms = int((time.time() - start) * 1000)
        results.append(IterationResult(
            iteration=i,
            model=config.primary_model,
            failure_mode=mode,
            success=success,
            latency_ms=latency_ms,
            error=error_msg,
            fail_open=client._last_fail_open,
        ))

    return results


# ─── Report ───────────────────────────────────────────────────────
def print_report(results: List[IterationResult]):
    print("\n" + "=" * 72)
    print("  CHAOS HARNESS REPORT (Python)")
    print("=" * 72)

    by_mode: Dict[str, List[IterationResult]] = {}
    for r in results:
        by_mode.setdefault(r.failure_mode, []).append(r)

    for mode, items in by_mode.items():
        successes = sum(1 for r in items if r.success)
        avg_latency = sum(r.latency_ms for r in items) / len(items)
        fail_opens = sum(1 for r in items if r.fail_open)

        print(f"\n  {mode.upper():16s} {len(items)} iterations")
        print(f"    Success:      {successes}/{len(items)}")
        print(f"    Avg latency:  {avg_latency:.0f}ms")
        if fail_opens > 0:
            print(f"    Fail-open:    {fail_opens}")
        for r in items:
            if r.error:
                print(f"    Error:        [iter {r.iteration}] {r.error}")

    total_success = sum(1 for r in results if r.success)
    total_fail_open = sum(1 for r in results if r.fail_open)
    print(f"\n  TOTAL: {total_success}/{len(results)} succeeded, {total_fail_open} fail-open")
    print("=" * 72 + "\n")


# ─── CLI entry ────────────────────────────────────────────────────
def main():
    parser = argparse.ArgumentParser(description="Chaos harness for Python SDK")
    parser.add_argument("--iterations", type=int, default=20)
    parser.add_argument("--latency-budget", type=int, default=5000)
    parser.add_argument("--max-cost-run", type=float, default=0.05)
    parser.add_argument("--max-cost-day", type=float, default=0.50)
    parser.add_argument("--base-url", default="https://api.openai.com")
    parser.add_argument("--api-key", default="")
    parser.add_argument("--host", default="http://localhost:3000")
    parser.add_argument("--run-id", default="")
    args = parser.parse_args()

    config = ChaosConfig(
        iterations=args.iterations,
        latency_budget_ms=args.latency_budget,
        max_cost_per_run=args.max_cost_run,
        max_cost_per_day=args.max_cost_day,
        base_url=args.base_url,
        api_key=args.api_key,
        host=args.host,
        run_id=args.run_id,
    )

    print(f"Running {config.iterations} chaos iterations...")
    print(f"Policy: enforce, latency={config.latency_budget_ms}ms, "
          f"cost/run=${config.max_cost_per_run}, cost/day=${config.max_cost_per_day}")
    print(f"Failure mix: {config.failure_mix}")

    results = run_chaos_harness(config)
    print_report(results)


if __name__ == "__main__":
    main()

File: packages/ts-sdk/harness/chaos.ts
/**
 * Chaos harness for TS SDK orchestration testing.
 *
 * Makes real (or realistically delayed) HTTP calls so timing and
 * cancellation behavior are tested under real network conditions.
 *
 * Usage:
 *   npx tsx packages/ts-sdk/harness/chaos.ts --iterations 20 --policy enforce
 */

import { FiveToOneClient, RuntimePolicy } from "../src/index";

// ─── Failure modes ────────────────────────────────────────────────
type FailureMode =
  | "success"
  | "timeout"       // sleep past latencyBudgetMs
  | "429"           // rate limit
  | "500"           // server error
  | "cost_overrun"; // fake large costUsd

interface ChaosConfig {
  iterations: number;
  failureMix: Record<FailureMode, number>; // 0-1 weights, normalized
  primaryModel: string;
  fallbackModels: string[];
  latencyBudgetMs?: number;
  maxCostPerRun?: number;
  maxCostPerDay?: number;
  baseUrl: string;  // real endpoint to hit for "success" calls
  apiKey: string;
  runId: string;
}

interface IterationResult {
  iteration: number;
  model: string;
  failureMode: FailureMode;
  success: boolean;
  latencyMs: number;
  error?: string;
  blocked?: boolean;
  fallbackReason?: string;
  failOpen?: boolean;
}

// ─── Sample config ────────────────────────────────────────────────
const DEFAULT_CONFIG: ChaosConfig = {
  iterations: 20,
  failureMix: {
    success: 0.4,
    timeout: 0.15,
    "429": 0.15,
    "500": 0.15,
    cost_overrun: 0.15,
  },
  primaryModel: "gpt-4o",
  fallbackModels: ["claude-sonnet-4-20250514", "gpt-4o-mini"],
  latencyBudgetMs: 5000,
  maxCostPerRun: 0.05,
  maxCostPerDay: 0.50,
  baseUrl: "https://api.openai.com",
  apiKey: process.env.OPENAI_API_KEY ?? "",
  runId: `chaos_${Date.now()}`,
};

// ─── Chaos call factory ───────────────────────────────────────────
function createChaosCall(config: ChaosConfig, mode: FailureMode) {
  return async (model: string, signal: AbortSignal): Promise<any> => {
    switch (mode) {
      case "success": {
        // Make a real HTTP call with abort support
        const controller = new AbortController();
        const timeout = setTimeout(() => controller.abort(), config.latencyBudgetMs ?? 30000);

        // Link external abort to internal
        signal.addEventListener("abort", () => controller.abort());

        try {
          const res = await fetch(`${config.baseUrl}/v1/chat/completions`, {
            method: "POST",
            headers: {
              "Content-Type": "application/json",
              "Authorization": `Bearer ${config.apiKey}`,
            },
            body: JSON.stringify({
              model,
              messages: [{ role: "user", content: "Say exactly: chaos test ok" }],
              max_tokens: 10,
            }),
            signal: controller.signal,
          });

          clearTimeout(timeout);
          if (!res.ok) throw new Error(`HTTP ${res.status}`);
          return await res.json();
        } catch (e) {
          clearTimeout(timeout);
          throw e;
        }
      }

      case "timeout": {
        // Sleep longer than latencyBudgetMs to trigger abort
        const sleepMs = (config.latencyBudgetMs ?? 5000) + 2000;
        await new Promise<void>((resolve, reject) => {
          const timer = setTimeout(resolve, sleepMs);
          signal.addEventListener("abort", () => {
            clearTimeout(timer);
            reject(new DOMException("The operation was aborted", "AbortError"));
          });
        });
        return { content: "should not reach here" };
      }

      case "429": {
        const err = new Error("429 Rate limit exceeded");
        (err as any).status = 429;
        throw err;
      }

      case "500": {
        const err = new Error("500 Internal Server Error");
        (err as any).status = 500;
        throw err;
      }

      case "cost_overrun": {
        // Return a large result that trips the cost ceiling
        return {
          content: "x".repeat(10000),
          _metadata: { estimatedCostUsd: 1.0 },
        };
      }

      default:
        throw new Error(`Unknown failure mode: ${mode}`);
    }
  };
}

// ─── Weighted random selection ────────────────────────────────────
function pickFailureMode(mix: Record<FailureMode, number>): FailureMode {
  const entries = Object.entries(mix) as [FailureMode, number][];
  const total = entries.reduce((s, [, w]) => s + w, 0);
  let r = Math.random() * total;
  for (const [mode, weight] of entries) {
    r -= weight;
    if (r <= 0) return mode;
  }
  return entries[entries.length - 1][0];
}

// ─── Run harness ──────────────────────────────────────────────────
async function runChaosHarness(config: ChaosConfig): Promise<IterationResult[]> {
  const client = new FiveToOneClient({
    apiKey: config.apiKey,
    host: "http://localhost:3000",  // point at local dev server
  });

  const results: IterationResult[] = [];

  for (let i = 0; i < config.iterations; i++) {
    const mode = pickFailureMode(config.failureMix);
    const allModels = [config.primaryModel, ...config.fallbackModels];
    const policy: RuntimePolicy = {
      enforcementMode: "enforce",
      fallbackChain: allModels,
      maxCostPerRun: config.maxCostPerRun,
      maxCostPerDay: config.maxCostPerDay,
      latencyBudgetMs: config.latencyBudgetMs,
      retryPolicy: {
        maxAttempts: 2,
        backoffMs: 500,
        backoffMultiplier: 2,
        retryableErrors: ["429", "5xx", "rate_limit", "timeout"],
      },
    };

    const calculateCost = mode === "cost_overrun"
      ? () => 1.0
      : undefined;

    const start = Date.now();
    let success = false;
    let error: string | undefined;

    try {
      await client.orchestrate({
        model: config.primaryModel,
        policy,
        call: createChaosCall(config, mode),
        input: { chaosIteration: i, failureMode: mode },
        calculateCost,
        runId: `${config.runId}_iter${i}`,
      });
      success = true;
    } catch (e) {
      error = e instanceof Error ? e.message : String(e);
    }

    results.push({
      iteration: i,
      model: config.primaryModel,
      failureMode: mode,
      success,
      latencyMs: Date.now() - start,
      error,
      failOpen: (client as any)._lastFailOpen,
    });
  }

  return results;
}

// ─── Report ───────────────────────────────────────────────────────
function printReport(results: IterationResult[]) {
  console.log("\n" + "═".repeat(72));
  console.log("  CHAOS HARNESS REPORT");
  console.log("═".repeat(72));

  const byMode = new Map<FailureMode, IterationResult[]>();
  for (const r of results) {
    const arr = byMode.get(r.failureMode) ?? [];
    arr.push(r);
    byMode.set(r.failureMode, arr);
  }

  for (const [mode, items] of byMode) {
    const successes = items.filter((r) => r.success).length;
    const avgLatency = items.reduce((s, r) => s + r.latencyMs, 0) / items.length;
    const failOpens = items.filter((r) => r.failOpen).length;

    console.log(`\n  ${mode.toUpperCase().padEnd(16)} ${items.length} iterations`);
    console.log(`    Success:      ${successes}/${items.length}`);
    console.log(`    Avg latency:  ${avgLatency.toFixed(0)}ms`);
    if (failOpens > 0) console.log(`    Fail-open:    ${failOpens}`);
    for (const r of items.filter((r) => r.error)) {
      console.log(`    Error:        [iter ${r.iteration}] ${r.error}`);
    }
  }

  const totalSuccess = results.filter((r) => r.success).length;
  const totalFailOpen = results.filter((r) => r.failOpen).length;
  console.log(`\n  TOTAL: ${totalSuccess}/${results.length} succeeded, ${totalFailOpen} fail-open`);
  console.log("═".repeat(72) + "\n");
}

// ─── CLI entry ────────────────────────────────────────────────────
const args = process.argv.slice(2);
const config = { ...DEFAULT_CONFIG };

for (let i = 2; i < args.length; i += 2) {
  switch (args[i]) {
    case "--iterations": config.iterations = Number(args[i + 1]); break;
    case "--latency-budget": config.latencyBudgetMs = Number(args[i + 1]); break;
    case "--max-cost-run": config.maxCostPerRun = Number(args[i + 1]); break;
    case "--max-cost-day": config.maxCostPerDay = Number(args[i + 1]); break;
    case "--base-url": config.baseUrl = args[i + 1]; break;
    case "--api-key": config.apiKey = args[i + 1]; break;
    case "--run-id": config.runId = args[i + 1]; break;
  }
}

console.log(`Running ${config.iterations} chaos iterations...`);
console.log(`Policy: enforce, latency=${config.latencyBudgetMs}ms, cost/run=$${config.maxCostPerRun}, cost/day=$${config.maxCostPerDay}`);
console.log(`Failure mix:`, config.failureMix);

runChaosHarness(config).then((results) => {
  printReport(results);
  process.exit(0);
}).catch((e) => {
  console.error("Harness failed:", e);
  process.exit(1);
});
